src/Cache/TemplateCache.php

```diff
@@ -25,14 +26,40 @@ class TemplateCache
      */
     protected ModuleHookLegacyRepository $hookModuleRepository;
 
+    /*
+     * @var Configuration
+     */
+    protected Configuration $configuration;
+
+    protected const CACHE_KEY = 'IS_MAINMENU_REQUEST_CACHE_KEY';
+
     public function __construct(
         \Is_mainmenu $module,
         \Context $context,
-        ModuleHookLegacyRepository $hookModuleRepository
+        ModuleHookLegacyRepository $hookModuleRepository,
+        Configuration $configuration
     ) {
         $this->module = $module;
         $this->context = $context;
         $this->hookModuleRepository = $hookModuleRepository;
+        $this->configuration = $configuration;
+    }
+
+    public function getContextBasedRequestCacheKey(): string
+    {
+        return md5($this->getRequestCacheKey() . '|' . $this->context->language->id . '|' . $this->context->shop->id . '|' . $this->context->customer->id_default_group);
```

Comment:
Food for thought:
- availability of categories to certain categories is something less used in the ecosystem, that's my POV, it'd be interesting to have cache key configurable 👍🏻